### PR TITLE
[extended-monitoring] migrating to go-lang #1311 

### PR DESCRIPTION
**Pull request by Horiodino**
## Description
The extended monitoring logic, originally implemented in Python, has been transitioned to the Go language
https://github.com/deckhouse/deckhouse/issues/1311#issuecomment-1837984193

## Why do we need it, and what problem does it solve?
The primary reason for transitioning the extended monitoring logic from Python to Go is due to the predominant use of the Go language within the project. By consolidating our monitoring logic into Go

## What is the expected result?
Upon transitioning the extended monitoring logic from Python to Go, the anticipated outcome is a robust and efficient monitoring system fully implemented in the Go language

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: extended-monitoring
type: feature 
summary: migrating to go-lang
impact_level: default 
```

### CODE REVIEW

**Comment by Horiodino** [comment]
i have tested but as the previous code was in python i dint understand vey well these are the outcomes:
Screenshot (52)
Screenshot (53)

**Comment by Horiodino** [comment]
> i have tested but as the previous code was in python i dint understand vey well these are the outcomes: Screenshot (52) Screenshot (53)

```apiVersion: v1
kind: Pod
metadata:
  name: nginx-pod
  labels:
    extended-monitoring.deckhouse.io/enabled: "true" 
spec:
  containers:
  - name: nginx-container
    image: nginx:latest  # NGINX image
    ports:
    - containerPort: 80
```

**Comment by Horiodino** [comment]
in the web its false cause i havent assigned the value as True , otherwise the code only exposes the objects having label ``` extended-monitoring.deckhouse.io/enabled: "true" ``` will be exposed

**Comment by Horiodino** [comment]
now what metrices I need to pull! I reviewed the python code, but I haven't worked with python so little bit hard to know what python code doing!

File: modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
+package main
```
Comment (alexey-gavrilov-flant) [inline modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.go:1]: you also need to rebuild the image.  https://github.com/Horiodino/deckhouse/blob/extended-monitoring-alerts-separately-%231311/modules/340-extended-monitoring/images/extended-monitoring-exporter/Dockerfile 
As an example, it is better to use this option https://github.com/Horiodino/deckhouse/blob/extended-monitoring-alerts-separately-%231311/modules/340-extended-monitoring/images/image-availability-exporter/Dockerfile

File: modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
```
Comment (alexey-gavrilov-flant) [inline modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.go:2]: and remove python version

File: modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+import (
```
Comment (alexey-gavrilov-flant) [inline modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.go:3]: You also need to correct the helm template for the golang variant  https://github.com/Horiodino/deckhouse/blob/extended-monitoring-alerts-separately-%231311/modules/340-extended- monitoring/templates/extended-monitoring-exporter/deployment.yaml#L64-L86

File: modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+	"context"
```
Comment (alexey-gavrilov-flant) [inline modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.go:4]: you need to take into account the changes in the documentation, but only after testing the new exporter variant. https://github.com/Horiodino/deckhouse/tree/extended-monitoring-alerts-separately-%231311/modules/340-extended-monitoring/docs

**Comment by Horiodino** [comment]
oh, ok 👍 but how's the info that is exposing, is it correct? or any changes!

> i have tested but as the previous code was in python i dint understand vey well these are the outcomes: Screenshot (52) Screenshot (53)

**Comment by alexey-gavrilov-flant** [comment]
the answer should be in prometeus metrics format https://prometheus.io/docs/concepts/metric_types/

**Comment by Horiodino** [comment]
todo : helm charts

**Comment by Horiodino** [comment]
https://github.com/deckhouse/deckhouse/issues/1311#issuecomment-1836520381

> @Horiodino why do you need a helm chart?

**Comment by alexey-gavrilov-flant** [comment]
@Horiodino I've discussed your PR with our architect.
You need to separate it into a Go transition 
and what needs to be done in the [problem](https://github.com/deckhouse/deckhouse/issues/1311). 
A PR that changes the logic and language of the exporter at the same time will be too hard to review.


[helm dir ](https://github.com/deckhouse/deckhouse/tree/main/modules/340-extended-monitoring/templates)

please replace your commits with the --signoff option.

**Comment by Horiodino** [comment]
ok 👍

File: modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.go [hunk 5]
```diff
@@ -0,0 +1,649 @@
+package extended_monitoring_exporter
```
[the remark this answers is not in the document]
Comment (alexey-gavrilov-flant) [reply modules/340-extended-monitoring/images/extended-monitoring-exporter/src/extended-monitoring.go:1]: https://github.com/Horiodino/deckhouse/pull/1

**Pull request closed without merging**